[OGUI-1580] add middleware for detector matching user role

pepijndik opened the pull request:
Middleware for checking if a user has a matching detector role
* Added User belongsToDetector

* DetectorOwnership middleware
* DetectorOwnershipMiddleware tests

#### I have JIRA issue created
- [x] branch and/or PR name(s) includes JIRA ID
- [x] issue has "Fix version" assigned
- [x] issue "Status" is set to "In review"
- [x] PR labels are selected
- [x] FLP integration tests were ran successful

Commit:
detectorLockMiddleware

Commit:
Adding the detectorOwnershipMiddleware

Commit:
Merge branch 'dev' into feature/pep/OGUI-1580/Addmiddlewarefordetectormatchinguserrole

Commit:
* Added User belongsToDetector
* DetectorOwnership middleware
* DetectorOwnershipMiddleware tests

Commit:
Fixed Eslint

Commit:
End of line enter

graduta commented on Control/lib/middleware/detectorOwnership.middleware.js:
```diff
@@ -0,0 +1,33 @@
+  
+  try {
+    const u = new User(username, name, personid, access);
+    console.log(`Checking locks for detector ${detectorId}`);
```
Hi, @pepijndik . We do not make use of `console` in this project but instead we use the `Logger` defined in the WebUI framework.
Nevertheless, in this case there is no need to log anything here in case of success.

EDIT: I have seen there are some `console.errors` in lockOwnership middleware and I confirm those are not correct and we should not be using it like this. I apologise if this is what tricked you into using it.

graduta commented on Control/lib/middleware/detectorOwnership.middleware.js:
```diff
@@ -0,0 +1,33 @@
+  }
+  
+  try {
+    const u = new User(username, name, personid, access);
```
Please refrain from using single-letter variables and instead use more descriptive ones

graduta commented on Control/lib/middleware/detectorOwnership.middleware.js:
```diff
@@ -0,0 +1,33 @@
+  
+    next(); // Proceed if lock ownership is verified
+  } catch (error) {
+    console.error(`Error checking locks:`, error);
```
The same as above, we do not use the `console` but instead the web-ui logger.

graduta commented on Control/lib/middleware/detectorOwnership.middleware.js:
```diff
@@ -0,0 +1,33 @@
+    const u = new User(username, name, personid, access);
+    console.log(`Checking locks for detector ${detectorId}`);
+    if (!u.belongsToDetector(detectorId)) {
+      return res
```
Rather than straight using the `res`, I would recommend you use the ` updateExpressResponseFromNativeError` with an `UnauthorizedError`

graduta commented on Control/lib/middleware/detectorOwnership.middleware.js:
```diff
@@ -0,0 +1,33 @@
+    next(); // Proceed if lock ownership is verified
+  } catch (error) {
+    console.error(`Error checking locks:`, error);
+    res.status(500).json({ message: 'Internal server error' });
```
Similar to above, please make use of ` updateExpressResponseFromNativeError`

graduta commented on Control/test/lib/middleware/mocha-detectorOwnership.middleware.test.js:
```diff
@@ -0,0 +1,100 @@
+    assert.ok(next.notCalled);
+  });
+
+  it('should integrate with an Express route and add detectorId', async () => {
```
This test does not follow the standard and moreover, it defines a new server just for the purpose of this middleware. In time, it can create confusion for developers. I would recommend to remove it.

I would encourage you to have a look at `api-get-locks.test.js` for example if you would like to see how to add tests to the API and make use of the `request` library to test the server API endpoint.
Of course, this would go into the api test directory

Commit:
Changed requested changes

Commit:
Code scanning, Removed unused imports

Commit:
Merge branch 'dev' into feature/pep/OGUI-1580/Addmiddlewarefordetectormatchinguserrole

graduta commented on Control/lib/middleware/detectorOwnership.middleware.js:
```diff
@@ -0,0 +1,33 @@
+  
+  try {
+    const user = new User(username, name, personid, access);
+    if (!user.belongsToDetector(detectorId)) {
```
Hi @pepijndik . Very good first step for the PR!
There is one thing that we did not discuss and I did not explain in the ticket. We should take into consideration the scenario in which:
- a user with a role higher than `Detector` is requesting the lock. In this scenario, they will not have the `det-<name>` permission but instead they will have a higher role. In this case, we should ignore the check of this middleware function

I am thinking before this `if` condition, you have another one in which you check that the role of the user is sufficient to go `next()`. Have a look at the function `isRoleSufficient`. You should check that the current role of user is at least `Role.GLOBAL`. If so, then you do not need to check if the user belongs to the detector

Commit:
DetectorOwnerShip Check Role

Check isRoleSufficient

Commit:
Merge branch 'dev' into feature/pep/OGUI-1580/Addmiddlewarefordetectormatchinguserrole

Commit:
Added Tests for RoleChecks

Detector ownership can have higher role than it's detector. Test this behavior

Commit:
Added ConvertStringToArray

Converts string into Array, with basic checks

Commit:
Merge remote-tracking branch 'origin/dev' into feature/pep/OGUI-1580/Addmiddlewarefordetectormatchinguserrole

Commit:
Removed unsused Import

Commit:
Merge branch 'dev' into feature/pep/OGUI-1580/Addmiddlewarefordetectormatchinguserrole

Commit:
Added stringToArray mocha tests

Commit:
Fixed Mocha stringToArray Test

Commit:
Fixed Api-lock tests

in URL detectorid is uppercase while in accesslist not so convert them to lower

graduta committed:
Merge branch 'dev' of github.com:AliceO2Group/WebUi into feature/pep/OGUI-1580/Addmiddlewarefordetectormatchinguserrole

graduta committed:
Update PR as per latest changes

graduta committed:
Fix incorrect name import

graduta committed:
Fix lint issues

graduta committed:
Improve testing of middleware

graduta committed:
Fix miss-spell in test

graduta committed:
Add missing new-line

graduta committed:
Merge branch 'dev' into feature/pep/OGUI-1580/Addmiddlewarefordetectormatchinguserrole

graduta committed:
Merge branch 'dev' of github.com:AliceO2Group/WebUi into feature/pep/OGUI-1580/Addmiddlewarefordetectormatchinguserrole

graduta committed:
Merge branch 'feature/pep/OGUI-1580/Addmiddlewarefordetectormatchinguserrole' of github.com:AliceO2Group/WebUi into feature/pep/OGUI-1580/Addmiddlewarefordetectormatchinguserrole

graduta committed:
Fix name move of file

graduta committed:
Update test error message
